File: src/main/java/de/tlmrgvf/drtd/utils/ReSampler.java
/*
 *
 * BSD 2-Clause License
 *
 * Copyright (c) 2020, Till Mayer
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package de.tlmrgvf.drtd.utils;

import de.tlmrgvf.drtd.dsp.component.biquad.BiquadFilter;
import de.tlmrgvf.drtd.dsp.component.biquad.GenericBiquadFilter;

public final class ReSampler {
    private final float targetSampleRatio;
    private final BiquadFilter lowPassFilter;
    private final boolean downSample;
    private float currentSampleRatio;
    private float average;
    private float averagedValues;
    private float currentSample;

    public ReSampler(float sourceSampleRate, float targetSampleRate) {
        targetSampleRatio = sourceSampleRate / targetSampleRate;
        currentSampleRatio = 0;
        if (sourceSampleRate > targetSampleRate) {
            lowPassFilter = new BiquadFilter(GenericBiquadFilter.Type.LOWPASS,
                    (int) sourceSampleRate,
                    targetSampleRate / 2);
        } else {
            lowPassFilter = null;
        }

        downSample = lowPassFilter != null;
    }

    public void processInputSample(float sample) {
        if (downSample) {
            sample = lowPassFilter.filterSample(sample);
            average += sample;
            ++averagedValues;
        }

        currentSample = sample;
        ++currentSampleRatio;
    }

    public Float readOutputSample() {
        if (currentSampleRatio >= 0) {
            currentSampleRatio -= targetSampleRatio;

            if (downSample) {
                if (averagedValues == 0)
                    return 0F;

                float value = average / averagedValues;
                averagedValues = 0;
                average = 0;
                return value;
            } else {
                return currentSample;
            }
        }

        return null;
    }
}

File: src/util/Resampler.cpp
/*
BSD 2-Clause License

Copyright (c) 2020, Till Mayer
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#include "Resampler.hpp"

using namespace Util;

Resampler::Resampler(float source_rate, float target_rate)
    : m_target_ratio(source_rate / target_rate)
    , m_down_sample(source_rate > target_rate) {
    assert(source_rate > 0 && target_rate > 0);
    if (m_down_sample)
        m_lowpass_filter = std::make_unique<Dsp::Biquad::Filter<float>>(Dsp::Biquad::Type::Lowpass, source_rate, target_rate / 2, Dsp::Biquad::invsqrt2);
}

void Resampler::process_input_sample(float sample) {
    if (m_down_sample) {
        sample = m_lowpass_filter->filter_sample(sample);
        m_average += sample;
        ++m_averaged_values;
    }

    m_current_sample = sample;
    ++m_current_ratio;
}

bool Resampler::read_output_sample(float& result) {
    if (m_current_ratio >= 0) {
        m_current_ratio -= m_target_ratio;

        if (m_down_sample) {
            if (m_averaged_values == 0) {
                result = 0;
                return true;
            }

            result = m_average / static_cast<float>(m_averaged_values);
            m_averaged_values = 0;
            m_average = 0;
            return true;
        } else {
            result = m_current_sample;
            return true;
        }
    }

    return false;
}
